### PR TITLE
feature/stock-tools-update-perfo

### PR DESCRIPTION
**Pull request by bm64**
Enhance stock tools with optimized update and redraw functionality.
- [x] - optional benchmark

### CODE REVIEW

File: ts/Stock/StockTools/StockToolbar.ts [hunk 1]
```diff
@@ -707,11 +726,67 @@ class Toolbar {
...
+                this.createButtons();
+            }
+
+            if (this.options.enabled === false) {
+                this.destroy();
+            }
+
+            if (defined(this.options.visible)) {
+                this.visible = this.options.visible;
+            }
```
Comment (pawellysy) [inline ts/Stock/StockTools/StockToolbar.ts:785]: There are a lot of if statements here. Would you mind splitting that into separate methods to improve code readability?

**Review by pawellysy — CHANGES_REQUESTED** [review]
Great job! 
Apart from the comment below, please try fixing the failing test. 
Also, could you add some tests? :)

File: ts/Stock/StockTools/StockToolbar.ts [hunk 2]
```diff
@@ -486,8 +502,28 @@ class Toolbar {
...
+        this.eventsToUnbind.push(
+            addEvent(this.showhideBtn, 'click', (): void => {
+                chart.update({
+                    stockTools: {
+                        gui: {
+                            visible: !self.visible,
+                            placed: true
+                        }
+                    }
+                });
```
Comment (pawellysy) [inline ts/Stock/StockTools/StockToolbar.ts:522]: StockTools has it's update method. Maybe we could use that to avoid some unnecessary logic?

PS: there is no `placed` property in the API. Should it be there?

**Review by pawellysy — CHANGES_REQUESTED** [review]
Great work! Just one small comment.

File: ts/Stock/StockTools/StockToolbar.ts [hunk 3]
```diff
@@ -486,8 +502,28 @@ class Toolbar {
             className: 'highcharts-stocktools-wrapper ' +
                 guiOptions.className + ' ' + bindingsClassName
         });
+
         container.appendChild(wrapper);
 
+        this.showhideBtn = createElement('div', {
```
Comment (kamil-musialowski) [inline ts/Stock/StockTools/StockToolbar.ts:508]: Is it `showhideBtn` on purpose, or should it be `showHideBtn`?

File: ts/Stock/StockTools/StockToolsGui.ts [hunk 4]
```diff
@@ -144,20 +143,29 @@ function onChartAfterGetContainer(
 function onChartBeforeRedraw(
     this: Chart
 ): void {
-    if (this.stockTools) {
+    onChartRedraw.call(this);
+    if (this.stockTools && this.stockTools.guiEnabled) {
```
Comment (kamil-musialowski) [inline ts/Stock/StockTools/StockToolsGui.ts:147]: Can we change it to optional chaining?
```suggestion
    if (this.stockTools?.guiEnabled) {
```

File: ts/Stock/StockTools/StockToolbar.ts [hunk 5]
```diff
@@ -472,12 +487,13 @@ class Toolbar {
...
      *
      */
-    public createHTML(): void {
+    public createContainer(): void {
         const chart = this.chart,
             guiOptions = this.options,
             container = chart.container,
             navigation = chart.options.navigation,
-            bindingsClassName = navigation && navigation.bindingsClassName;
+            bindingsClassName = navigation && navigation.bindingsClassName,
```
Comment (kamil-musialowski) [inline ts/Stock/StockTools/StockToolbar.ts:495]:
```suggestion
            bindingsClassName = navigation?.bindingsClassName,
```

File: ts/Stock/StockTools/StockToolbar.ts [hunk 6]
```diff
@@ -649,6 +668,20 @@ class Toolbar {
         }
     }
 
+    /**
+     * Remove highcharts-active class from button.
+     * @private
+     */
+    public removeButtonActiveClass(
```
Comment (kamil-musialowski) [inline ts/Stock/StockTools/StockToolbar.ts:675]: I'm afraid that 4fd095a6fa7a99d7c279fe748321becbb437430f removed the usage of this method, and we just have a function, but it's not used anywhere.

So we either need to remove this function, or revert the changes from the commit that removed the usage.

**Review by kamil-musialowski — CHANGES_REQUESTED** [review]
Just a few comments, nothing major tbh. Great job overall! 👏

File: samples/unit-tests/stock-tools/gui/demo.js [hunk 7]
```diff
@@ -229,3 +229,54 @@ QUnit.test(
...
+    assert.ok(
+        chart.stockTools.toolbar.classList.contains('updatedToolbarClassName'),
+        'Toolbar should have updated class name.'
+    );
+
+    assert.strictEqual(
+        chart.stockTools.toolbar.children.length,
+        2,
+        'Stock Tools should have correct number of buttons after update.'
+    );
```
Comment (pawelfus) [inline samples/unit-tests/stock-tools/gui/demo.js:281]: Could you also add one more assert? After disabling GUI, check if elements are correctly hidden/destroyed

File: ts/Stock/StockTools/StockToolbar.ts [hunk 8]
```diff
@@ -707,11 +740,92 @@ class Toolbar {
...
+    }
+
+    /**
+     * Hadles the change of the `enabled` option.
+     * @private
+     */
+    private handleGuiEnabledChange(): void {
+        if (this.options.enabled === false) {
+            this.destroy();
+            this.guiEnabled = false;
```
Comment (pawelfus) [inline ts/Stock/StockTools/StockToolbar.ts:769]: unnecessary line

**Review by pawelfus — CHANGES_REQUESTED** [review]
Nice! See my suggestions below

**Review by DJTechnoo — APPROVED** [review]
Amazing work!

File: ts/Stock/StockTools/StockToolsGui.ts [hunk 9]
```diff
@@ -145,38 +144,46 @@ function onChartBeforeRedraw(
     this: Chart
 ): void {
     if (this.stockTools) {
-        const optionsChart = this.options.chart as ChartOptions;
-        const listWrapper = this.stockTools.listWrapper,
-            offsetWidth = listWrapper && (
+        this.stockTools.redraw();
+
+        if (this.stockTools.guiEnabled) {
```
Comment (pawelfus) [inline ts/Stock/StockTools/StockToolsGui.ts:149]: Q: What do you think about extracting the whole code below to a function, eg `setOffset()` or `calculateOffset()`?

**Review by pawelfus — CHANGES_REQUESTED** [review]
Merge with master to get rid of failing tests. Other than that it looks good, nice! 🎉

File: ts/Stock/StockTools/StockToolsDefaults.ts [hunk 10]
```diff
@@ -1260,9 +1260,9 @@ const stockTools: StockToolsOptions = {
             }
         },
         /**
-         * Wether the stock tools toolbar is visible.
+         * Whether the stock tools toolbar is visible.
          *
-         * @type {boolean}
```
Comment (kamil-musialowski) [inline ts/Stock/StockTools/StockToolsDefaults.ts]: Removing that `@type` was on purpose? 🤔

**Review by kamil-musialowski — APPROVED** [review]
All good imo! Please verify that one little comment 👍

**Review by pawellysy — APPROVED** [review]

File: ts/Stock/StockTools/StockToolsDefaults.ts (hunk 10 above)
Comment (bm64) [reply ts/Stock/StockTools/StockToolsDefaults.ts]: Yes https://github.com/highcharts/highcharts/pull/21204#discussion_r1629683499

**Review by pawelfus — APPROVED** [review]
Thanks!

**Review by TorsteinHonsi — APPROVED** [review]
Thanks!
